rfc/rfc-79/rfc-90.md

```diff
@@ -0,0 +1,310 @@
+is just to make sure, an inflight plan won't stay on the timeline for an unbounded amount of time but also won't be
+likely to be prematurely cleaned up by clean before it has a chance to be executed.
+
+An additional config "cancellation-policy" will be added to the table service plan to indicate when it is eligible to be
+cancelled and aborted by `CLEAN` operation. This policy can be a threshold of hours or instants on timeline, where if
+that # of hours/instants have elapsed since the plan was scheduled(w/o being executed), any call to `CLEAN` operation
+will also cancel and abort the instant. This policy should be configured by the worker scheduling a cancellable table
+service plan, based on the amount of time they expect the plan to remain on the timeline before being picked up for
+execution. For example, if the plan is expected to have its execution deferred to a few hours later, then the
+cancellation-policy should be lenient in allowing the plan to remain many hours on the timeline before it could be
+deemed eligible for abortion. Note that this cancellation policy is not used to infer whether a clustering plan is
+currently being executed - similar to how concurrent ingestion writes are safely rolled back by lazy clean policy, A
+cancellable clustering plan can only be aborted once it is confirmed that an ongoing writer is no longer progressing it
+(which the aforementioned execute_abort API already enforces). Note that the cancellation-policy is only required to be honored by `CLEAN`
+
+To implement this optional feature, the clean operation can use both of these APIs in order to cancel any incomplete plans that have
+satisfied their cancellation policy or already been requested for cancellation. This is analagous to how clean already schedules and
+executes the rollback of any failed ingestion writes. Specifically, clean will now perform the following steps
+
+1. Iterate through each inflight cancellable instant, for each instant that has met their cancellation policy, call
+   request_cancel if not already has an request for cancellation.
+2. Iterate through each instant in /.cancel and call execute_abort. Note that the request_cancel and execute_abort APIs
+   are responsible for ensuring that already-committed instants aren't cancelled and that ongoing writers aren't
+   concurrently having their files cleaned up by a writer doing cancellation.
+
+
+
+### Summarizing proposed internal timeline/instant format changes
+
+Taking together all changes to .hoodie & instant files structure discussed above, this RFC will apply the following
+changes
+
+- The `cancellable` boolean flag and `cancellation-policy` predicates will be added as optional serialized values to the
+  table service metadata schema (tentatively just for clustering plans)
+- The list of possible instant transitions for table services will now include `.aborted`
```

Comment:
If we are introducing a new action and state, I think we should also callout the interplay with different query types, especially incremental queries. Many prod users use 2-stage incremental pull model. I am not sure how `.aborted` will affect. Will instants with such a state be simply ignored?

Reply:
Oh I thought that incremental readers (incremental queries or Hudi streamer jobs that read new commits from an upstream dataset and write to a "silver"/"gold" etc dataset downstream) only look at committed instants. Since I thought that similar to inflight or rolled back instants, .aborted instants would be ignored. But let me researc a bit more into 
> 2-stage incremental pull mode
since it seems I might be misunderstanding there

Reply:
Actually sorry can you share an example reference to "2-stage incremental pull mode". I was under the impression that an inflight commit on source table will be ignored by a streamer/incremental query job until it is committed. But based on your comment I am getting the impression that there might be some optimization I am unaware of in HUDI streamer/incremental query where data files for an inflight write are optimistically/prematurely read (under the assumption that the inflight write will be committed by the time the incremental query is finished).